### PR TITLE
Lock for Bounty

### LINKED ISSUE #385: Lock period for bounty
Need more exploration. Please test a few scenarios where a bounty hunter can be a source of exploit.

### PR DESCRIPTION
**Pull request by abhishekvispute**
Fixes #385 

Hi Guys,
This PR adds lock for bounty 
Please check if the execution is correct and if the implementation can be improved.

The process is as per this 
Per slash, we return a bountyId to bountyHunter
then there is lock period
once its over, using that id, bountyHunter can redeem their bounty

_Have not used the existing lock of unstake/withdraw as there some extra param are involved like staker.address, resetLock, commission and all_

CC : @hrishikeshio @SkandaBhat @SamAg19

### CODE REVIEW

**Comment by SkandaBhat** [comment]
Please update with master branch.

File: contracts/Core/StakeManager.sol [hunk 1]
```diff
@@ -293,22 +293,44 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
...
+
+        require(msg.sender == bountyLocks[bountyId].bountyHunter, "Incorrect Caller");
+        require(bountyLocks[bountyId].redeemAfter <= epoch, "Redeem epoch not reached");
+        delete bountyLocks[bountyId];
+
+        //reward half the amount to bounty hunter
+        //please note that since slashing is a critical part of consensus algorithm,
+        //the following transfers are not `reuquire`d. even if the transfers fail, the slashing
+        //tx should complete.
+        razor.transfer(msg.sender, bounty);
```
Comment (hrishikeshio) [inline contracts/Core/StakeManager.sol:333]: require?

Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol]: done https://github.com/razor-network/contracts/pull/411/commits/b3e1ae9bd7eb6892be5124b6071eff34a2dcde09

**Comment by abhishekvispute** [comment]
> Please update with master branch.

done, gasCompare is available now

**Review by hrishikeshio — APPROVED** [review]

**Comment by abhishekvispute** [comment]
Hi Guys,
Due to recent additions on master in StakeManager , like event for resetLock and enums for stake change.
The contract size has overshoot the size limit with this PR.
we will have to pull it back by some optimization or dividing the contract.

**Review by SkandaBhat — DISMISSED** [review]

**Comment by hrishikeshio** [comment]
> Hi Guys,
> Due to recent additions on master in StakeManager , like event for resetLock and enums for stake change.
> The contract size has overshoot the size limit with this PR.
> we will have to pull it back by some optimization or dividing the contract.

you can try reducing optimizer runs to slightly reduce contract size

**Comment by abhishekvispute** [comment]
> you can try reducing optimizer runs to slightly reduce contract size

Hi Guys, @razor-network/engineering 
Have reduced opt runs by half.
Please check gasCompare

<html>
<body>


VoteManager | commit | (+)0.12 | 159143 | 158954 | (+)189.00
-- | -- | -- | -- | -- | --
VoteManager | reveal | (+)0.05 | 176065 | 175974 | (+)91.00
<html>
<body>


BlockManager | propose | (+)0.03 | 259897 | 259807 | (+)90.00
-- | -- | -- | -- | -- | --



</body>
</html>

</body>
</html>

Let me know if we should keep it or should move bounty out in sep contract.

**Comment by abhishekvispute** [comment]
Hi Guys,
Have reduced optimizer runs for stakeManager by half, for others it remains the same

**Review by hrishikeshio — APPROVED** [review]

**Review by SamAg19 — DISMISSED** [review]

**Pull request merged**